### PR TITLE
Get-DbaSqlBuildReference, allow online updates

### PR DESCRIPTION
**Pull request by niphlod**
Changes proposed in this pull request:
 - added a new Update switch to fetch the index from online assets

### CODE REVIEW

File: functions/Get-DbaSqlBuildReference.ps1 [hunk 1]
```diff
@@ -105,13 +116,26 @@ Function Get-DbaSqlBuildReference {
...
 				$data_time = Get-Date $data_content.LastUpdated
 				
+				$offline_time = $module_time
 				if ($module_time -gt $data_time) {
 					Copy-Item -Path $orig_idxfile -Destination $writable_idxfile -Force -ErrorAction Stop
 					$result = $module_content
-				}
-				
-				else {
+				} else {
```
Comment (FriedrichWeinmann) [inline functions/Get-DbaSqlBuildReference.ps1]: Was there another deterioration in our style guideline?
If not, please refrain from placing else on the same line as the closing brace :shudder:

File: functions/Get-DbaSqlBuildReference.ps1 [hunk 2]
```diff
@@ -105,13 +116,26 @@ Function Get-DbaSqlBuildReference {
...
+					$offline_time = $data_time
+				}
+				# If Update is passed, try to fetch from online resource and store into the writeable
+				if ($Update) {
+					$WebContent = Get-DbaSqlBuildReferenceIndexOnline -Silent $Silent
+					if ($null -ne $WebContent) {
+						$webdata_content = $WebContent.Content | ConvertFrom-Json
+						$webdata_time = Get-Date $webdata_content.LastUpdated
+						if ($webdata_time -gt $offline_time) {
+							Write-Message -Level Output -Message "Index updated correctly, last update on: $(Get-Date -Date $webdata_time -f s), was $(Get-Date -Date $offline_time -f s)"
```
Comment (FriedrichWeinmann) [inline functions/Get-DbaSqlBuildReference.ps1]: Don't use aliases - even for parameters - in a script

File: functions/Get-DbaSqlBuildReference.ps1 [hunk 3]
```diff
@@ -122,12 +146,35 @@ Function Get-DbaSqlBuildReference {
...
 			$result.Data | Select-Object @{ Name = "VersionObject"; Expression = { [version]$_.Version } }, *
 		}
 		
+		function Get-DbaSqlBuildReferenceIndexOnline {
+			[CmdletBinding()]
+			Param (
+				[bool]
+				$Silent
+			)
+			$url = 'https://sqlcollaborative.github.io/assets/dbatools-buildref-index.json'
```
Comment (FriedrichWeinmann) [inline functions/Get-DbaSqlBuildReference.ps1]: Should be a configuration entry. Information should not be hardcoded. For example, large organizations could provide an internal path and thus allow machines without online access to update.
I'd recommend creating a new configuration file "assets", then place the config there and call it with something like this:
`Get-DbaConfigValue -Name 'assets.sqlbuildreference'`

File: functions/Get-DbaSqlBuildReference.ps1 (hunk 2 above)
Comment (niphlod) [inline functions/Get-DbaSqlBuildReference.ps1]: huh ... aliases ??

Comment (FriedrichWeinmann) [inline functions/Get-DbaSqlBuildReference.ps1]: `-f` instead of `-Format`

File: functions/Get-DbaSqlBuildReference.ps1 (hunk 3 above)
Comment (niphlod) [inline functions/Get-DbaSqlBuildReference.ps1]: good thinking. But I don't want to disobey 'cause "disobedience shall be answered by the wrath of Fred." :-P I'm missing how to code the "handler" part for it.

File: functions/Get-DbaSqlBuildReference.ps1 (hunk 1 above)
Comment (niphlod) [inline functions/Get-DbaSqlBuildReference.ps1]: will adhere but I lost where this is formalized ^_^ pretty recently edited files never had this explicit guideline

File: functions/Get-DbaSqlBuildReference.ps1 (hunk 2 above)
Comment (niphlod) [inline functions/Get-DbaSqlBuildReference.ps1]: dang! Fixed.

File: functions/Get-DbaSqlBuildReference.ps1 (hunk 3 above)
Comment (FriedrichWeinmann) [inline functions/Get-DbaSqlBuildReference.ps1]: Handler is optional. It's mostly input validation. You could try to convert the input into an uri, in order to determine whether it's legit, but you can also just skip validation if you want to.

**Comment by niphlod** [comment]
moved $url to configuration, too. Thanks for the review

File: internal/configurations/assets.ps1 [hunk 4]
```diff
@@ -0,0 +1,6 @@
+<#
+This is designed for all online assets you need configurations for.
+#>
+
+# The default path where dbatools stores persistent data
+Set-DbaConfig -Name 'assets.sqlbuildreference' -Value 'https://sqlcollaborative.github.io/assets/dbatools-buildref-index.json' -Default -DisableHandler -Description "The url where dbatools fetches the up to date builreference index"
```
Comment (FriedrichWeinmann) [inline internal/configurations/assets.ps1]: Typo: `Builreference`
You may also want to include the product the build reference is for in the description (just to be 100% blunt and direct)

Comment (niphlod) [inline internal/configurations/assets.ps1]: ok

File: functions/Get-DbaSqlBuildReference.ps1 [hunk 5]
```diff
@@ -25,6 +25,11 @@ Function Get-DbaSqlBuildReference {
 		Returns information about a build identified by  "12.00.4502" (which is SQL 2014 with SP1 and CU11)
 	
 	.EXAMPLE
+		Get-DbaSqlBuildReference -Build "12.00.4502" -Update
```
Comment (PowerDBAKlaas) [inline functions/Get-DbaSqlBuildReference.ps1]: Can you add 'update' in the .PARAMETER section, please?
Maybe some extensive explanation what happens when using it; does it overwrite a file?

Comment (niphlod) [inline functions/Get-DbaSqlBuildReference.ps1]: done

**Pull request merged**